src/spox/opset/ai/onnx/v17.py

```diff
@@ -1780,16 +1780,20 @@ class Outputs(BaseOutputs):
         v_final_and_scan_outputs: Sequence[_VarInfo]
 
     def infer_output_types(self, input_prop_values: PropDict) -> dict[str, Type]:
-        output_types = super().infer_output_types({})
+        output_types = super().infer_output_types(input_prop_values)
+        output_names = list(self.outputs.get_var_infos())
 
         body = self.attrs.body.value
-        n = len(body.requested_arguments) - 2
 
-        carried_names = list(self.outputs.get_var_infos())[:n]
-        carried_types = [v.type for v in list(body.requested_results.values())[1:][:n]]
+        initial_types = [v.type for v in list(body.requested_arguments)[2:]]
+        carried_types = [v.type for v in list(body.requested_results.values())[1:]]
 
-        for name, typ in zip(carried_names, carried_types):
-            output_types[name] = typ
+        is_constant_shape = all(
+            i_typ == c_typ for i_typ, c_typ in zip(initial_types, carried_types)
+        )
```

Comment:
```suggestion
        shape_unchanged_between_iterations = all(
            i_typ == c_typ for i_typ, c_typ in zip(initial_types, carried_types)
        )
```

nitpick:  it's obvious enough from the context but "constant" is a bit of an overloaded term at this point.